Modify register screen to use material design (and be accessible)

mdinino opened the pull request:
### Description

[MA-2698](https://openedx.atlassian.net/browse/MA-2698)

-Replace TextViews in registration screen with TextInputLayout
-Do not use TextInputLayout error handling. It is inconsistent with other fields, continue to use test fields to display errors (on TextInputLayout and Spinner fields).
- Replace spinners with native spinner (but have the underline style).
- "Sign up with" and "or sign up with email" has a value that is an appropriate consorts ratio.
- "Hide optional field" and "Show optional field" is blue like a hyperlink.
- Delete "instruction field" in layout and json file.
- Set the unselected values appropriately.
- "Tell us why you're interesdted in edX" starts off at one line and grows to a maximum of 5 lines.

### Testing
- [ ] Unit, integration, acceptance tests as appropriate
 Attempt to register with values that aren't set and make sure that error message shows up appropriately.
- Turn on talkback mode and ensure that user can navigate through all the fields and labels appropriately.
- Register a new user (this is the happy path).

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @bguertin 
- [x] Code review: @miankhalid 
- [ ] Code review: @1zaman 
- [x] UX review: @lizcohen
- [ ] FYI @andy-armstrong

mdinino commented:
New registration screen: 
![screen shot 2016-08-29 at 2 01 07 pm](https://cloud.githubusercontent.com/assets/18707535/18061541/14d3ca92-6df1-11e6-9045-7c6aee61eaeb.png)

mdinino commented:
Old registration screen: 
![screen shot 2016-08-29 at 2 04 32 pm](https://cloud.githubusercontent.com/assets/18707535/18061654/8bdb5b64-6df1-11e6-80d0-65e7d893d445.png)

mdinino commented:
@lizcohen gave me a verbal :+1

bguertin commented:
This is a problem in the existing design, but it would be nice if the Spinners had labels, especially after values are selected. Sadly, there is no built-in mechanism for this, but it's quite strange to have spinners with no labels.

lizcohen commented:
what spinners are we referring to?

On Mon, Aug 29, 2016 at 4:31 PM, Brian @ edX <notifications@github.com>
wrote:

> This is a problem in the existing design, but it would be nice if the
> Spinners had labels, especially after values are selected. Sadly, there is
> no built-in mechanism for this, but it's quite strange to have spinners
> with no labels.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-app-android/pull/815#issuecomment-243246091>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ASyECO3S58XLLxPpoZ9EW1lPoakDs-sqks5qk0GZgaJpZM4Jvup0>
> .
>



-- 
Liz Cohen
UX Designer

bguertin commented:
> what spinners are we referring to?

Country, Gender, Birth Year.

Willing to defer, since that problem is not introduced in this PR.

BenjiLee commented:
👍  for screenshots.

Can we add more space on top of the FB/G+ buttons?

mdinino commented:
@bguertin et. all. I did find custom solutions floating around online that has labels on spinners.

Just as one example, theres this: https://github.com/rey5137/material

I didn't know how much time I should focus on getting that to work. But if we want, that is something I can do.

mdinino commented:
@BenjiLee I addressed your comment. Here is a new screen shot.
![screen shot 2016-08-30 at 3 49 23 pm](https://cloud.githubusercontent.com/assets/18707535/18104536/8737a878-6ec9-11e6-9c23-0e16b80903f8.png)

mdinino commented:
@BenjiLee I added space to the FB/G+ buttons. New screenshot 
![screen shot 2016-08-30 at 3 49 23 pm](https://cloud.githubusercontent.com/assets/18707535/18131813/405010fa-6f62-11e6-8846-d998c961404b.png)

mdinino commented:
@bguertin @1zaman @miankhalid It's it's not too much trouble, I'd like to get this merged soon. It is very closely tied with MA-2763 (which I am also working on). In fact, I think it's best if I don't create a formal PR for MA-2763 until this one is merged.

miankhalid commented on VideoLocker/res/layout/activity_register.xml:
```diff
@@ -36,18 +36,18 @@
-                    android:textColor="@color/light_gray"
-                    android:textSize="14sp" />
+                    android:textColor="@color/grey_text"
+                    android:textSize="14sp"/>
```
`@dimen/edx_small` can be used instead of this hardcoded textSize.

mdinino replied:
Addressed

miankhalid commented on VideoLocker/res/layout/activity_register.xml:
```diff
@@ -57,19 +57,19 @@
                     android:text="@string/or_sign_up_with_email"
-                    android:textColor="@color/light_gray"
+                    android:textColor="@color/grey_text"
                     android:textSize="14sp"
```
@dimen/edx_small can be used instead of this hardcoded textSize.

mdinino replied:
Addressed

miankhalid commented on VideoLocker/res/layout/activity_register.xml:
```diff
@@ -86,33 +86,19 @@
                     android:focusable="true"
-                    android:background="@drawable/selectable_box_overlay"
                     android:text="@string/show_optional_text"
-                    android:textColor="@color/grey_hide_optional_text"
```
This color can be deleted now from `colors.xml`.

mdinino replied:
Addressed

miankhalid commented on VideoLocker/res/layout/activity_register.xml:
```diff
@@ -86,33 +86,19 @@
                     android:layout_marginLeft="12dp"
                     android:layout_marginStart="12dp"
                     android:layout_marginRight="12dp"
                     android:layout_marginEnd="12dp"
```
Can we standardise these hardcoded values to `dimens.xml` alternatives ?

mdinino replied:
Addressed

miankhalid commented on VideoLocker/res/layout/activity_register.xml:
```diff
@@ -86,33 +86,19 @@
-                    android:background="@drawable/selectable_box_overlay"
                     android:text="@string/show_optional_text"
-                    android:textColor="@color/grey_hide_optional_text"
                     android:textSize="14sp"
```
`@dimen/edx_small` can be used instead of this hardcoded textSize.

mdinino replied:
Addressed

miankhalid commented on VideoLocker/res/layout/registration_spinner_item.xml:
```diff
@@ -0,0 +1,13 @@
+    android:textAlignment="inherit"
+    android:fontFamily="sans-serif"
+    android:paddingTop="16dp"
+    android:textSize="17sp" />
```
`@dimen/large` or `@dimen/base` can be used instead of this hardcoded textSize.
p.s. Git loves newlines at end of file.

mdinino replied:
Addressed

miankhalid commented on VideoLocker/res/layout/registration_spinner_item.xml:
```diff
@@ -0,0 +1,13 @@
+    android:layout_height="wrap_content"
+    android:ellipsize="marquee"
+    android:textAlignment="inherit"
+    android:fontFamily="sans-serif"
```
Is this really needed ?

mdinino replied:
Which part are you referring to? The main reason I added in this XML was for android:fontFamily="sans-serif" (i.e. Roboto Regular). The other attributes makes sure that we don't lose anything when switching from the default item view to this one.

miankhalid replied:
Why do we need to use Roboto Regular here ? we are using OpenSans-Regular throughout the app.

mdinino replied:
We want RoboRegular so that the spinner matches with the TextInputLayout (defaults to RobotoRegular). We want that to match the TextInputLayout in the Login activity.

miankhalid commented on VideoLocker/res/layout/registration_spinner_item.xml:
```diff
@@ -0,0 +1,13 @@
+<TextView xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    style="?android:attr/spinnerItemStyle"
+    tools:targetApi="17"
```
Is this really needed ? I don't see an attribute that's targeted for API 17. Also, if we remove this line, we can get rid of the namespace inclusion as well i.e. `xmlns:tools="http://schemas.android.com/tools"`.

mdinino replied:
Addressed

miankhalid commented on VideoLocker/res/layout/view_register_spinner.xml:
```diff
@@ -3,32 +3,20 @@
-        android:paddingEnd="10dp"
+        android:textColorHint="@color/grey_text"
         android:focusable="true"
+        android:paddingLeft="-10dp"
```
A negative padding ?

mdinino replied:
Yeah, this is to get the text to line up with the EditText items above and below the spinners. There is precedent in activity_login.xml

miankhalid replied:
Ok, if they are necessary, maybe move them to `dimens.xml`.

miankhalid commented on VideoLocker/res/layout/view_register_spinner.xml:
```diff
@@ -3,32 +3,20 @@
+        android:textColorHint="@color/grey_text"
         android:focusable="true"
+        android:paddingLeft="-10dp"
         tools:targetApi="17"/>
```
Not needed.

mdinino replied:
Addressed. Although Android Studio made me add both paddingStart and paddingLeft.

miankhalid commented on VideoLocker/res/layout/view_register_spinner.xml:
```diff
@@ -3,32 +3,20 @@
         android:id="@+id/input_spinner_error"
         style="@style/registration_error_message"
-        android:visibility="visible"
-        tools:text="error text here"/>
```
Having this turns out to be beneficial while viewing the layout in Android Studio's layout viewer, so, it shouldn't be removed.

mdinino replied:
Addressed.

miankhalid commented on VideoLocker/res/values/text_styles.xml:
```diff
@@ -260,7 +260,9 @@
-        <item name="android:padding">5dp</item>
+        <item name="android:paddingTop">5dp</item>
+        <item name="android:paddingLeft">2dp</item>
+        <item name="android:paddingRight">2dp</item>
```
`dimens.xml` resources maybe ?

miankhalid commented:
Done with the 1st pass.

mdinino commented:
@miankhalid I have addressed your comments from the first pass.

mdinino commented:
@miankhalid @1zaman @bguertin What is the likelihood that I can have this merged by tomorrow (the end of the sprint)?

miankhalid commented on VideoLocker/res/layout/activity_register.xml:
```diff
@@ -91,13 +91,10 @@
-                    android:layout_marginStart="12dp"
-                    android:layout_marginRight="12dp"
-                    android:layout_marginEnd="12dp"
+                    android:layout_margin="@dimen/edx_x_small"
```
This will add a `marginTop` as well, do we need that ?

mdinino replied:
Good catch. We don't. I fixed it.

miankhalid commented on VideoLocker/res/layout/activity_register.xml:
```diff
@@ -91,13 +91,10 @@
                     android:text="@string/show_optional_text"
-                    android:textSize="14sp"
+                    android:textSize="@dimen/edx_small"
                     tools:targetApi="17" />
```
If we plan to not have `marginEnd` and `marginStart` attributes, then this attribute can be deleted.

mdinino replied:
Addressed.

miankhalid commented on VideoLocker/res/layout/registration_spinner_item.xml:
```diff
@@ -1,13 +1,11 @@
-    android:textSize="17sp" />
\ No newline at end of file
+    android:paddingTop="@dimen/edx_base"
+    android:textSize="@dimen/edx_base" />
```
Maybe a newline at the end of this layout file, cuz Git apparently loves them 😉

mdinino replied:
done.

miankhalid commented on VideoLocker/res/values/dimens.xml:
```diff
@@ -8,6 +8,10 @@
 
+    <!-- Registration screen -->
+    <dimen name="registration_below_fields_vertical">5dp</dimen>
+    <dimen name="registration_below_fields_horizontal">2dp</dimen>
```
Names are not that hip. How about these ?
- `registration_field_vertical_padding`
- `registration_field_horizontal_padding`

mdinino replied:
Addressed

miankhalid commented:
Other than some minor nits this LGTM 👍

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/module/registration/view/RegistrationOptionSpinner.java:
```diff
@@ -1,15 +1,20 @@
 
+import org.edx.mobile.R;
 import org.edx.mobile.module.registration.model.RegistrationOption;
-import org.edx.mobile.view.custom.CustomSelectView;
```
I think `CustomSelectView`is now unused so it can be deleted

mdinino replied:
Deleted

mdinino commented:
@bguertin @1zaman Polite code review nudge

BenjiLee commented:
@mdinino Lahore is out for the next 3 days. I'll take a look at some of these in the meantime.

BenjiLee commented:
When registration fails, talkback does not announce anything.

BenjiLee commented:
When in talkback mode, when selecting a spinner, focus does not remain on the spinner after selecting a spinner item.

BenjiLee commented on VideoLocker/res/layout/registration_spinner_item.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
```
nit: Is this spinner registration specific, or can we make this the general edX spinner layout?

mdinino replied:
It is registration specific. It is formatted to look well on the registrations screen (in others words, it can't be easily abstracted to spinners in general...at least not now)

BenjiLee replied:
Wrong module

BenjiLee replied:
Wrong module

BenjiLee commented on VideoLocker/res/layout/registration_spinner_item.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
+<TextView xmlns:android="http://schemas.android.com/apk/res/android"
+    style="?android:attr/spinnerItemStyle"
+    android:singleLine="true"
```
Use maxLines instead.

BenjiLee commented on VideoLocker/res/layout/view_register_edit_text.xml:
```diff
@@ -1,29 +1,24 @@
-        android:visibility="visible"
-        tools:text="error text here"/>
+        <android.support.design.widget.TextInputEditText
+            android:id="@+id/tilEt"
```
Please rename this ID.

BenjiLee commented on VideoLocker/res/layout/view_register_edit_text.xml:
```diff
@@ -1,29 +1,24 @@
 
-    <EditText
+    <android.support.design.widget.TextInputLayout
+        android:id="@+id/til"
```
Please rename this ID

BenjiLee commented on VideoLocker/res/layout/view_register_spinner.xml:
```diff
@@ -3,32 +3,22 @@
+        android:textColorHint="@color/grey_text"
         android:focusable="true"
+        android:paddingStart="-10dp"
+        android:paddingLeft="-10dp"
```
Use dimen resources

BenjiLee commented on VideoLocker/res/layout/view_register_spinner.xml:
```diff
@@ -3,32 +3,22 @@
     xmlns:tools="http://schemas.android.com/tools"
```
nit: Could we also make this a general spinner?

mdinino replied:
@BenjiLee This is not abstract enough to be general spinner. There is a custom error layout here.

BenjiLee commented:
Done with first pass.

mdinino commented:
@BenjiLee Re: When in talkback mode, when selecting a spinner, focus does not remain on the spinner after selecting a spinner item.

I am going to push back and suggest that this is out of scoope. I agree that it is annoying, but it seems to be the default Android behavior (I am using a default Android spinner and adapter afterall).

BenjiLee commented:
Given that this is an a11y ticket to update the registration UI to make it a11y friendly, I would argue that focus control is within scope of this ticket. Just because something is native, doesn't mean that it's not a problem. Either way, you should check with a11y.

mdinino commented:
@BenjiLee I addressed your comments. I am going to spend some time figuring out what, if anything, can be done about the spinner focus. I will timebox it to an hour or so. If nothing is apparent, I will ask Mark what he wants to do.

mdinino commented:
@BenjiLee I don't see much about it that I can do. I also created a sample app to confirm that this is, indeed default Android behavior. That app worked the same. I also saw the same behavior in 2 established apps (one of them being YouTube). I couldn't fin Mark, but I talked to Scott. He said that if this is what the established pattern is, then it's fine.

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/module/registration/view/RegistrationCheckBoxView.java:
```diff
@@ -82,24 +74,27 @@ public View getView() {
 
     @Override
-    public void handleError(String error) {
+    public void handleError(String error, boolean requestAccessibility) {
```
Is there any way to avoid adding a specific boolean to this function?

mdinino replied:
@BenjiLee We need some way for the activity to tell the control weather or not to request accessibility focus on error. If we don't then the accessibility focus will just go to every error. From a user perspective, that would mean that the last error would have the focus.

I think it's a better user experience for only the first to have the focus. So a conditional value is needed. What do you think?

mdinino replied:
I suppose, I could make a function that passes false as a default. But nothing would call that function. thoughts?

BenjiLee replied:
Lets try to avoid passing booleans through multiple methods.

BenjiLee commented:
@mdinino Again, I want to re emphasize that just because something is native does not mean it is accessible. For example, EditText is not a11y friendly but it is a native widget.

 In this case, if it's too hard to fix the behaviour then let's not worry about it for now.

mdinino commented:
@BenjiLee I agree with you in theory, but if Scott is saying not to make a custom control to get around default behavior then I'm not sure what else to do.

BenjiLee commented:
> In this case, if it's too hard to fix the behaviour then let's not worry about it for now.

mdinino commented:
@BenjiLee I found a way to avoid passing around boolean values. Ready for re-review.

BenjiLee commented:
Maybe we should have a popup dialog on submission if there are errors on the submission form. There is no feedback when submitting. The focus remains on the "Create my account" button.

mdinino commented:
@BenjiLee I added in a pop-up in the case of an error:
<img src="https://cloud.githubusercontent.com/assets/18707535/18688389/357a5f8c-7f52-11e6-9daa-0aa4159db561.png" width="250" height="500"/>

mdinino commented:
@BenjiLee I'm ready for a re-review.

mdinino commented:
@catong What do you think about the string on the popup dialog? I like my suggestion, but I feel like we can do better.

catong commented on VideoLocker/res/values/strings.xml:
```diff
@@ -505,6 +505,9 @@
     <string name="try_again">Try again</string>
+    <!-- Registration error -->
+    <string name="registration_error_title">Registration Error</string>
+    <string name="registration_error_message">Please verify the information that was entered.</string>
```
@mdinino Just a small tweak. How about "Please verify the information that you entered."

mdinino replied:
I like it.

catong commented on VideoLocker/assets/config/registration_form.json:
```diff
@@ -35,7 +33,6 @@
         "errorMessages": {},
         "placeholder": "",
         "defaultValue": "",
```
Yep, just checking, because I believe that that help text was added deliberately to address the pain point of learners not understanding which name was used where, and not registering with a name they wanted on their certs. It is also important to communicate that usernames once chosen, can't be changed. I defer to @marcotuts if this was a Product call, I just was not aware of the decision. Thanks!

catong replied:
Were other options discussed for other ways to show this help? e.g. in placeholder text within the fields?

mdinino replied:
We talked about this a few weeks ago, but if I recall @lizcohen and @scottrish were part of that conversation. I think they thought that the extra information wasn't necessary.

@lizcohen? What are your thoughts?

marcotuts replied:
Thanks for the reply here @catong. We should either: 
1.) merge this but not update the apps until we resolve the inclusion of these support labels.
2.) or include the text labels within this PR (whether it is a separate ticket or not is a separate issue)

BenjiLee commented:
Waiting on help text before review.

lizcohen commented:
I think we can keep the help text off. If it proves to have problems/cause confusion we can add back in.

mdinino commented:
@lizcohen I like it better this way too. It's cleaner.

lizcohen commented:
As for the registration error screen, I don't think using the text "Please verify the information that you entered" works because we're not really asking the user to verify the info, there was something wrong, so they need to fix something. How about something like "Please update highlighted field(s) and try again" cc: @catong

mdinino commented:
@catong Do you agree with Liz's string proposal?

catong commented:
Yes, good point, @lizcohen that the string should be about fixing issues rather than verifying. "Update the highlighted fields and try again." works, as long as we are highlighting fields.

I'm not clear on the resolution for the help issue - @marcotuts were you agreeing that including guidance is preferred? I'd really like to help users proactively wherever we can, rather than wait for support issues to come in and then react. I think learners confusion about the way that we use full name vs username in the system is a known one. 

I just want to clarify that I'm not necessarily suggesting simply restoring what we had previously (Help text below the field?) - we could consider something different such as making the guidance part of the example text within the field - but defer to @lizcohen on the display decision.

BenjiLee commented:
> "Update the highlighted fields and try again." 

Note: Whatever we decide on, lets make this message a11y friend.

mdinino commented:
@catong @lizcohen, I am trying to decide whether to have two separate strings:

One if there is only one error (e.g. Update the highlighted field and try again.) 
And another if there is another error (e.g. Update the highlighted fields and try again).

This is an easy change from an implementation perspective, but it will add another entry in strings.xml.

What are you thoughts? If you are indifferent, I will implement it with two strings.

catong commented:
@mdinino @lizcohen I personally think it's fine to have a single string that refers to "fields" (plural), even if there is just one field highlighted. If it makes it easier, we could use "any" instead: "Update any highlighted fields and try again."

BenjiLee commented:
> @catong @lizcohen, I am trying to decide whether to have two separate strings:
> 
> One if there is only one error (e.g. Update the highlighted field and try again.) 
> And another if there is another error (e.g. Update the highlighted fields and try again).
> 
> This is an easy change from an implementation perspective, but it will add another entry in strings.xml.
> 
> What are you thoughts? If you are indifferent, I will implement it with two strings.
> 

Just use string plurals.

mdinino commented:
@marcotuts Re: registration error screen: We previously didn't have an alert dialog for registration errors. We are adding it for accessibility reasons.

According to this Wiki: https://openedx.atlassian.net/wiki/display/MA/Error+Handling
all alert dialogs should block until the error is either dismissed by the user or the back button is pressed.

mdinino commented:
@lizcohen How do you feel about @catong's suggestion of modifying the following labels?
"Full name" -> "Full name: used in certificates"
"Public username" -> "Public username: cannot be changed"

lizcohen commented:
@mdinino @catong  How about removing the colons and making the text that would have been after the colon 80% smaller. Also, I think it should say "used on certificates" rather than "used in certificates"

catong commented:
@lizcohen Sounds OK to me, although what point size are we going to arrive at if 80% smaller - any issues with accessibility?
So: "Full name" "Used on certificates"
and "Public username" "Cannot be changed"

@marcotuts @cptvitamin Your thoughts?

mdinino commented:
@BenjiLee @catong @lizcohen I updated the error string, take a look.

Also, I just realized something else about the string labels (e.g. "Full name" "Used on certificates", etc). We are pulling that information from the server and that API call is also used on the web.

I can get rid of the description field (e.g. "This name appears on the certificate you earn" ) on the UI by just not showing it, but I can't modify it without affecting the web. And in previous discussions, that is something that we don't wan to touch.

How do you feel about closing out this PR and investigating the possible need for "descriptions" in another ticket?

BenjiLee commented:
> I updated the error string, take a look.

If you squash your commits, its hard to see the updates. Try not to squash if you are making small changes that need to be visible. 

> Also, I just realized something else about the string labels (e.g. "Full name" "Used on certificates", etc). We are pulling that information from the server and that API call is also used on the web.

Would like to emphasize that any string modification we make here to registration.json will be moot when we start fetching the registration form from the server.

cptvitamin commented:
WCAG actually does not specify a min font size. What they do require is that the user can change the font size to suit their preferences. So, when a user changes the font size preferences of their OS (in the Settings), it should also scale the text in native apps like edX. The spec actually specifies that the text should be able to scale 200%. I think, by using Material Design here, users can scale the font size by adjusting their display settings.

Where we have more control over the font size used,  I just like to apply common sense here.  You want people to read the text right?  Can you read the text?  Can your grandmother with cataracts?  Channel your inner grandmother.  Humans are living longer, healthier lives. Let's design interfaces that everyone can read and understand without squinting and holding their devices up to their faces ;)

mdinino commented:
@BenjiLee Agreed. We are fetching from the server, so we don't have much leverage for text, font size, etc.

BenjiLee commented:
> @BenjiLee Agreed. We are fetching from the server, so we don't have much leverage for text, font size, etc.

We have full control over all of those. What we don't have control over is the content returned.

mdinino commented:
Ah, so you're saying use the content from the server as a trigger, but then reference a localized string? Is that what I should do? Should I go forward with that?

lizcohen commented:
> I updated the error string, take a look.
@mdinino where can I see this?

mdinino commented:
@lizcohen This was in a screenshot above.
<img src="https://cloud.githubusercontent.com/assets/18707535/18788535/38534122-8175-11e6-8e8e-616b0d99941c.png" width="250" height="500"/>

BenjiLee commented:
> > "Update the highlighted fields and try again." 
> 
> Note: Whatever we decide on, lets make this message a11y friend.

Bump.

mdinino commented:
@catong @lizcohen Should we go back to ""Please verify the information that you entered."? I think that one is best for visible and non visible users alike.

BenjiLee commented:
> Ah, so you're saying use the content from the server as a trigger, but then reference a localized string? Is that what I should do? Should I go forward with that?

Not sure what you mean by "use the content from the server as a trigger". Also, let's not worry about localization in this PR since it is out of scope.

BenjiLee commented:
Sounds like don't want to exclude the description. What if we added an (?) icon on the right side of the layout that has the description.

<img src="http://www.technotalkative.com/wp-content/uploads/2011/09/android-setError.png"  height="250"/>

BenjiLee commented:
Could you make the "Sign up with" and "or "Sign up with email" more consistent with the login screen? 

<img src="https://cloud.githubusercontent.com/assets/7101723/18791977/ccae6ede-8182-11e6-94bf-6fd74c1d70d1.png"  height="400"/>

mdinino commented:
@lizcohen @BenjiLee @andy-armstrong Screenshot implementation of what we discussed on Friday:
![screen shot 2016-09-23 at 5 32 38 pm](https://cloud.githubusercontent.com/assets/18707535/18801933/dabd8024-81b3-11e6-84a7-3c575053bd03.png)

mdinino commented:
Due to an error on my part, @lizcohen put this comment in the wrong PR. I am putting it here
@mdinino Can we move the helper text and error messages up a bit closer to the line? Like so: 
<img width="513" alt="screen shot 2016-09-24 at 8 13 35 am" src="https://cloud.githubusercontent.com/assets/19694600/18808391/dc27b274-822e-11e6-97e5-bff31f538f77.png">

mdinino commented:
Liz gave me a verbal 👍 .

@BenjiLee Ready for a re-review.

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/view/custom/DividerWithTextView.java:
```diff
@@ -0,0 +1,31 @@
+package org.edx.mobile.view.custom;
```
Wrong module

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/view/dialog/AlertDialogFragment.java:
```diff
@@ -12,38 +13,82 @@
  * Note: This is a very simple implementation that only shows an {@link AlertDialog} with a given
  * message and an OK button without any listener.
- *
+ * <p/>
```
Is this intentional?

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/view/RegistrationTextAreaView.java:
```diff
@@ -9,18 +9,19 @@
+        mTextInputEditText.setGravity(Gravity.START | Gravity.TOP);
     }
-}
+}
```
newline

BenjiLee commented on OpenEdXMobile/res/values/dimens.xml:
```diff
@@ -8,6 +8,12 @@
     <!-- Per Material Design Guidelines -->
     <dimen name="action_bar_icon_size">24dp</dimen>
 
+    <!-- Registration screen -->
```
We have a section for registration. Please combine.

BenjiLee commented:
This is out of scope. I was looking at the iOS app. The Google+ and FB button have content descriptions to include "Register with" so when you move focus on it it will say "Register with Google". Up to you to include this.

BenjiLee commented:
Done with pass. UI looks good.

mdinino commented:
@BenjiLee I addressed your recent round of comments. I want to get this ticket landed to start on What's New so I'm going to call the content description item out of scope for now. Good find though.

catong commented:
@mdinino @BenjiLee I just saw this - in a meeting but need to make some edits to the Helper text. Will get you comments ASAP

BenjiLee commented:
There should be no VideoLocker folder.

mdinino commented:
@BenjiLee There should not being anything associated with the VideoLocker folder in this PR.

BenjiLee commented on OpenEdXMobile/res/layout/view_register_spinner.xml:
```diff
@@ -3,32 +3,30 @@
-        android:paddingEnd="10dp"
+        android:textColorHint="@color/grey_text"
         android:focusable="true"
+        android:paddingStart="@dimen/registration_spinner_horizontal_padding"
```
Define end and right padding.

mdinino replied:
@BenjiLee This was just to putt the passing to the left to line it up with the text input layout. If you notice the values are negative numbers. No right/end definition makes sense here.

BenjiLee replied:
Its for RTL support.

mdinino replied:
Padding doesn't affect RTL support. As the right/end padding needs no definition here, I think its' best to leave it out

BenjiLee replied:
Could you explain how padding doesn't affect RTL support?

BenjiLee replied:
Nevermind. I figured it out. Padding **in this case** doesn't affect RTL since we don't care about symmetry.

BenjiLee commented on OpenEdXMobile/res/values/dimens.xml:
```diff
@@ -166,6 +170,7 @@
     <!-- Stray dimensions -->
     <dimen name="notification_settings_margin">12dp</dimen>
     <dimen name="fullscreen_message_max_width">240dp</dimen>
+    <dimen name="divider_with_text_margin">12dp</dimen>
```
There are other social divider dimens. Please organize.

mdinino replied:
Those "divider" margins specific to the padding above and below the divers, not within the divider itself. I'll move it, but I think the necessity to point this out is debateable.

BenjiLee commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -469,9 +469,9 @@
     <!-- In Sign up screen -->
-    <string name="sign_up_with">Sign up with</string>
+    <string name="sign_up_with">Register with</string>
     <!-- In Sign up screen -->
```
Update docstring

mdinino replied:
What's docstring? I looked in confluence and couldn't find anything. Is it somewhere else?

mdinino replied:
Ah, you mean the comment. Got it.

BenjiLee commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -469,9 +469,9 @@
     <!-- Prompt leading user to create account screen -->
     <string name="sign_up_and_learn">Sign up and start learning</string>
     <!-- In Sign up screen -->
```
update docstring

BenjiLee commented on OpenEdXMobile/res/values/text_styles.xml:
```diff
@@ -250,17 +250,10 @@
-        <item name="android:gravity" tools:targetApi="17">left|start|center_vertical</item>
-        <item name="android:textSize">11sp</item>
-        <item name="android:padding">5dp</item>
+    <style name="registration_instruction_message" parent="registration_error_message">
```
This style should have a docstring since it's a style for a custom implementation of textinputlayout.

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/view/RegistrationAgreementView.java:
```diff
@@ -17,7 +18,8 @@
     protected RegistrationFormField mField;
     private View mView;
     protected TextView mInputView;
```
private

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/view/RegistrationAgreementView.java:
```diff
@@ -17,7 +18,8 @@
     protected RegistrationFormField mField;
```
private

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/view/RegistrationCheckBoxView.java:
```diff
@@ -14,34 +15,29 @@
-    private RegistrationFormField mField;
     private View mView;
+    private RegistrationFormField mField;
     protected CheckBox mInputView;
```
private

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/view/RegistrationEditTextView.java:
```diff
@@ -15,79 +17,67 @@
+    protected TextInputLayout mTextInputLayout;
+    protected TextInputEditText mTextInputEditText;
+    protected TextView mInstructionsTextView;
+    protected TextView mErrorTextView;
```
Any reason why these are all protected?

mdinino replied:
They are references by subclasses. They're protected in  master too (and are accessed directly)

BenjiLee replied:
That's a bit odd but I guess we should leave the refactoring out for now. Just because something is in master doesn't mean it's correct.

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/view/RegistrationOptionSpinner.java:
```diff
@@ -23,23 +28,51 @@ public RegistrationOptionSpinner(Context context, AttributeSet attrs, int defSty
+        if (adapter != null) {
+            ret = (getAdapterPosition(value) >= 0);
+        }
+        return ret;
```
@mdinino Could you use more descriptive variable names?

https://www.cs.umd.edu/~clin/MoreJava/Intro/var-descr.html

BenjiLee replied:
```
public boolean hasValue(@Nullable String value){
    if (adapter != null) {
       return (getAdapterPosition(value) >= 0);
    }
    return false;
}
```

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/view/RegistrationOptionSpinner.java:
```diff
@@ -23,23 +28,51 @@ public RegistrationOptionSpinner(Context context, AttributeSet attrs, int defSty
+            ret = selected.getValue();
         }
-        return false;
+        return ret;
```
@mdinino Could you use more descriptive variable names?

https://www.cs.umd.edu/~clin/MoreJava/Intro/var-descr.html

BenjiLee replied:
```
RegistrationOption selected = (RegistrationOption)getSelectedItem();
if (selected != null) {
    return selected.getValue();
}
return null;
```

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/view/RegistrationOptionSpinner.java:
```diff
@@ -23,23 +28,51 @@ public RegistrationOptionSpinner(Context context, AttributeSet attrs, int defSty
+                }
             }
         }
+        return ret;
```
@mdinino Could you use more descriptive variable names?

https://www.cs.umd.edu/~clin/MoreJava/Intro/var-descr.html

mdinino replied:
It's the name of the variable that will be returned. Again, I will change it, but is this really necessary to hold up a PR?

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/view/RegistrationSelectView.java:
```diff
@@ -18,23 +19,21 @@
     private RegistrationFormField mField;
     private View mView;
     protected RegistrationOptionSpinner mInputView;
```
private

BenjiLee commented on OpenEdXMobile/src/main/java/org/edx/mobile/module/registration/view/RegistrationTextAreaView.java:
```diff
@@ -9,18 +9,19 @@
     // Number of lines for TextArea
-    private static final int LINES = 5;
+    private static final int INIT_LINES = 1;
+    private static final int MAX_LINES = 5;
```
Now that the font is significantly bigger in textinputlayout, we should also increase the max lines.

BenjiLee commented:
Done with pass.

BenjiLee commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -469,9 +469,9 @@
     <!-- Prompt leading user to create account screen -->
     <string name="sign_up_and_learn">Sign up and start learning</string>
```
@catong Should this change to "Register and start learning?"

mdinino commented:
@BenjiLee  Ready for another review

BenjiLee commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -467,11 +467,11 @@
-    <string name="sign_up_with">Sign up with</string>
-    <!-- In Sign up screen -->
-    <string name="or_sign_up_with_email">or sign up with email</string>
+    <string name="register_and_learn">Register and start learning</string>
```
@mdinino You need to refactor the string with the IDE. If you only change the text here, the references to this string won't be updated.

BenjiLee commented on OpenEdXMobile/res/values/dimens.xml:
```diff
@@ -65,11 +65,16 @@
     <dimen name="login_logo_margin_bottom">15dp</dimen>
     <dimen name="login_social_divider_margin_top">20dp</dimen>
     <dimen name="login_social_divider_margin_bottom">15dp</dimen>
+    <dimen name="divider_with_text_margin">12dp</dimen>
```
This is in both the Register and Login activity. This shouldn't be here.

mdinino replied:
@BenjiLee I told you yesterday that I didn't think it belonged with the other divider dimens. But I fixed it to appease you. Now I put it iust below the only 2 instances that say "dimen" and you don't like it. Where would you like me to put it?

BenjiLee replied:
@mdinino If you didn't think it belonged with the other divider you should have kept it there and said why?

BenjiLee commented:
@mdinino done with pass.

mdinino commented:
I brought it up. Moving forward, how can we land this?

On Fri, Sep 30, 2016 at 11:13 AM Christopher Lee <notifications@github.com>
wrote:

> *@BenjiLee* commented on this pull request.
> ------------------------------
>
> In OpenEdXMobile/res/values/dimens.xml
> <https://github.com/edx/edx-app-android/pull/815>:
>
> > @@ -65,11 +65,16 @@
>      <dimen name="login_logo_margin_bottom">15dp</dimen>
>      <dimen name="login_social_divider_margin_top">20dp</dimen>
>      <dimen name="login_social_divider_margin_bottom">15dp</dimen>
> +    <dimen name="divider_with_text_margin">12dp</dimen>
>
> @mdinino <https://github.com/mdinino> If you didn't think it belonged
> with the other divider you should have kept it there and said why?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-app-android/pull/815>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR10TxHaZIu7EnXgJ_0DQtlMjicWuXjBks5qvScJgaJpZM4Jvup0>
> .
>

BenjiLee commented:
You made the change on your own accord and then brought it up. I would move it back?

mdinino commented:
BenjiLee a day ago
There are other social divider dimens. Please organize.
       @mdinino
mdinino 21 hours ago
Those "divider" margins specific to the padding above and below the divers, not within the divider itself. I'll move it, but I think the necessity to point this out is debateable.

BenjiLee commented:
Exactly. So why did you move it against your own explanation?

BenjiLee commented:
@mdinino 

> Moving forward, how can we land this?

- Move the dimension to the place you explained it should be.
- Rename the string properly that is failing tests.

mdinino commented:
In the event of an RTL language, the passing will be swapped. Not defining
a padding will mean that the default is used. It’s safe.

On Fri, Sep 30, 2016 at 4:07 PM Christopher Lee <notifications@github.com>
wrote:

> *@BenjiLee* commented on this pull request.
> ------------------------------
>
> In OpenEdXMobile/res/layout/view_register_spinner.xml
> <https://github.com/edx/edx-app-android/pull/815>:
>
> >          android:focusable="true"
> +        android:paddingStart="@dimen/registration_spinner_horizontal_padding"
>
> Could you explain how padding doesn't affect RTL support?
>
> —
> You are receiving this because you modified the open/close state.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-app-android/pull/815>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR10T9FYFJmtfoUC7G1lp8P5QO0II0sMks5qvWv7gaJpZM4Jvup0>
> .
>

mdinino commented:
*padding will be swapped

On Fri, Sep 30, 2016 at 4:08 PM Marc Dinino <mdinino@edx.org> wrote:

> In the event of an RTL language, the passing will be swapped. Not defining
> a padding will mean that the default is used. It’s safe.
>
> On Fri, Sep 30, 2016 at 4:07 PM Christopher Lee <notifications@github.com>
> wrote:
>
>> *@BenjiLee* commented on this pull request.
>> ------------------------------
>>
>> In OpenEdXMobile/res/layout/view_register_spinner.xml
>> <https://github.com/edx/edx-app-android/pull/815>:
>>
>> >          android:focusable="true"
>> +        android:paddingStart="@dimen/registration_spinner_horizontal_padding"
>>
>> Could you explain how padding doesn't affect RTL support?
>>
>> —
>> You are receiving this because you modified the open/close state.
>>
>>
>> Reply to this email directly, view it on GitHub
>> <https://github.com/edx/edx-app-android/pull/815>, or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AR10T9FYFJmtfoUC7G1lp8P5QO0II0sMks5qvWv7gaJpZM4Jvup0>
>> .
>>
>

BenjiLee commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -467,11 +467,10 @@
-    <string name="register_with">Register with</string>
-    <!-- In register screen -->
-    <string name="or_register_with_email">Or register with email</string>
+    <string name="sign_up_and_learn">Sign up and start learning</string>
```
This is the string we wanted to revert because it is out of scope. This is currently correct.

BenjiLee commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -467,11 +467,10 @@
-    <!-- In register screen -->
-    <string name="or_register_with_email">Or register with email</string>
+    <string name="sign_up_and_learn">Sign up and start learning</string>
+    <!-- In Sign Up screen -->
```
`Sign up` -> `Register`

BenjiLee commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -467,11 +467,10 @@
-    <string name="or_register_with_email">Or register with email</string>
+    <string name="sign_up_and_learn">Sign up and start learning</string>
+    <!-- In Sign Up screen -->
+    <string name="sign_up_with">Sign up with</string>
```
This within the scope of this PR and needs to be reverted to what was originally discussed. 
`Sign up` -> `Register`

BenjiLee commented on OpenEdXMobile/res/values/strings.xml:
```diff
@@ -467,11 +467,10 @@
+    <string name="sign_up_and_learn">Sign up and start learning</string>
+    <!-- In Sign Up screen -->
+    <string name="sign_up_with">Sign up with</string>
+    <string name="or_sign_up_with_email">or sign up with email</string>
```
This within the scope of this PR and needs to be reverted to what was originally discussed. 
Sign up -> Register

BenjiLee commented:
👍  merge
